core/src/duplicate_repair_status.rs

```diff
@@ -239,20 +239,69 @@ impl AncestorRequestStatus {
             }
             let our_frozen_hash = blockstore.get_bank_hash(*ancestor_slot);
             if let Some(our_frozen_hash) = our_frozen_hash {
-                if earliest_erroring_ancestor.is_some() && our_frozen_hash == *agreed_upon_hash {
-                    // It's impossible have a different version of an earlier ancestor, but
-                    // then also have the same version of a later ancestor.
-                    info!("{} mismatches then matches", self.requested_mismatched_slot);
-                    return DuplicateAncestorDecision::InvalidSample;
-                } else if our_frozen_hash != *agreed_upon_hash
-                    && earliest_erroring_ancestor.is_none()
-                {
-                    earliest_erroring_ancestor = Some((
-                        agreed_response.len() - i - 1,
-                        DuplicateAncestorDecision::EarliestMismatchFound(
-                            DuplicateSlotRepairStatus::default(),
-                        ),
-                    ));
+                match (
+                    &earliest_erroring_ancestor,
+                    our_frozen_hash == *agreed_upon_hash,
+                ) {
+                    (
+                        Some((mismatch_i, DuplicateAncestorDecision::EarliestMismatchFound(_))),
+                        true,
+                    ) => {
+                        // It's impossible have a different version of an earlier ancestor, but
+                        // then also have the same version of a later ancestor.
+                        let (mismatch_slot, mismatch_agreed_upon_hash) =
+                            agreed_response[*mismatch_i];
+                        let mismatch_our_frozen_hash = blockstore.get_bank_hash(mismatch_slot);
+                        info!(
+                            "When processing the ancestor sample for {}, there was a mismatch
+                            for {mismatch_slot}: we had frozen hash {:?} and the cluster agreed upon
+                            {mismatch_agreed_upon_hash}. However for a later ancestor {ancestor_slot}
+                            we have agreement on {our_frozen_hash} as the bank hash. This should never
+                            be possible, something is wrong or the cluster sample is invalid.
+                            Rejecting and queuing the ancestor hashes request for retry",
+                            self.requested_mismatched_slot,
+                            mismatch_our_frozen_hash
+                        );
+                        return DuplicateAncestorDecision::InvalidSample;
+                    }
+                    (
+                        Some((mismatch_i, DuplicateAncestorDecision::EarliestAncestorNotFrozen(_))),
```

Comment:
Just checks if we found an unfrozen mismatch earlier. Could disambiguate it even further if needed lmk
- Catch frozen hash matches, then random unfrozen block, then frozen hash matches again (shouldn't happen even in snapshot)
- Pass earliest block available in blockstore and ignore mismatches before then (could get tricky if block is getting dump & repaired by state machine on startup)

Reply:
> Catch frozen hash matches, then random unfrozen block, then frozen hash matches again (shouldn't happen even in snapshot)

I think this fix is fine for now. we could probably detect this by keeping track of the number of frozen hash matches and checking if it's `> 0` when we detect a frozen hash after an unfrozne block